File: backend-aspnet/PetshopApi/Models/Pedido.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace PetshopApi.Models;

public enum StatusPedido
{
    Pendente,
    Confirmado,
    Processando,
    Enviado,
    Entregue,
    Cancelado
}

[Table("pedidos")]
public class Pedido
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public long Id { get; set; }

    [Required(ErrorMessage = "Data do pedido é obrigatória")]
    [Column("data_pedido")]
    public DateTime DataPedido { get; set; } = DateTime.Now;

    [Required(ErrorMessage = "Valor total é obrigatório")]
    [Range(0.01, double.MaxValue, ErrorMessage = "Valor deve ser maior que zero")]
    [Column("valor_total")]
    public double ValorTotal { get; set; }

    [Column("status")]
    public StatusPedido Status { get; set; } = StatusPedido.Pendente;

    [StringLength(50)]
    [Column("forma_pagamento")]
    public string? FormaPagamento { get; set; } // credito, debito, pix, boleto

    [StringLength(1000)]
    [Column("observacoes")]
    public string? Observacoes { get; set; }

    // Relacionamentos
    [Required]
    [Column("cliente_id")]
    public long ClienteId { get; set; }

    [ForeignKey("ClienteId")]
    public Cliente? Cliente { get; set; }

    public ICollection<ItemPedido> Itens { get; set; } = new List<ItemPedido>();

    // Métodos auxiliares
    public void CalcularValorTotal()
    {
        ValorTotal = Itens.Sum(item => item.Subtotal);
    }

    public void AdicionarItem(ItemPedido item)
    {
        Itens.Add(item);
        item.Pedido = this;
        CalcularValorTotal();
    }

    public void RemoverItem(ItemPedido item)
    {
        Itens.Remove(item);
        item.Pedido = null;
        CalcularValorTotal();
    }
}

File: functions-java/petshop-functions-shared/src/main/java/com/petshop/functions/shared/model/Pedido.java
package com.petshop.functions.shared.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "pedidos")
public class Pedido {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull(message = "Data do pedido é obrigatória")
    @Column(name = "data_pedido", nullable = false)
    private LocalDateTime dataPedido;

    @NotNull(message = "Valor total é obrigatório")
    @DecimalMin(value = "0.0", inclusive = false, message = "Valor deve ser maior que zero")
    @Column(name = "valor_total", nullable = false)
    private Double valorTotal;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private StatusPedido status = StatusPedido.PENDENTE;

    @Column(name = "forma_pagamento", length = 50)
    private String formaPagamento; // credito, debito, pix, boleto

    @Column(length = 1000)
    private String observacoes;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cliente_id", nullable = false)
    private Cliente cliente;

    @OneToMany(mappedBy = "pedido", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ItemPedido> itens = new ArrayList<>();

    // Enum para status do pedido
    public enum StatusPedido {
        PENDENTE,
        CONFIRMADO,
        PROCESSANDO,
        ENVIADO,
        ENTREGUE,
        CANCELADO
    }

    // Construtores
    public Pedido() {
        this.dataPedido = LocalDateTime.now();
    }

    public Pedido(Cliente cliente) {
        this();
        this.cliente = cliente;
    }

    // Getters e Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public LocalDateTime getDataPedido() {
        return dataPedido;
    }

    public void setDataPedido(LocalDateTime dataPedido) {
        this.dataPedido = dataPedido;
    }

    public Double getValorTotal() {
        return valorTotal;
    }

    public void setValorTotal(Double valorTotal) {
        this.valorTotal = valorTotal;
    }

    public StatusPedido getStatus() {
        return status;
    }

    public void setStatus(StatusPedido status) {
        this.status = status;
    }

    public String getFormaPagamento() {
        return formaPagamento;
    }

    public void setFormaPagamento(String formaPagamento) {
        this.formaPagamento = formaPagamento;
    }

    public String getObservacoes() {
        return observacoes;
    }

    public void setObservacoes(String observacoes) {
        this.observacoes = observacoes;
    }

    public Cliente getCliente() {
        return cliente;
    }

    public void setCliente(Cliente cliente) {
        this.cliente = cliente;
    }

    public List<ItemPedido> getItens() {
        return itens;
    }

    public void setItens(List<ItemPedido> itens) {
        this.itens = itens;
    }

    // Métodos auxiliares
    public void adicionarItem(ItemPedido item) {
        itens.add(item);
        item.setPedido(this);
        calcularValorTotal();
    }

    public void removerItem(ItemPedido item) {
        itens.remove(item);
        item.setPedido(null);
        calcularValorTotal();
    }

    public void calcularValorTotal() {
        this.valorTotal = itens.stream()
                .mapToDouble(ItemPedido::getSubtotal)
                .sum();
    }

    @PrePersist
    @PreUpdate
    protected void onSave() {
        if (this.dataPedido == null) {
            this.dataPedido = LocalDateTime.now();
        }
        calcularValorTotal();
    }
}
